File: src/actions/registry.go
package actions

import (
	"fmt"
	"strings"
	"enva/libs"
	"enva/services"
)

var actionRegistry = make(map[string]ActionFactory)

// RegisterAction registers an action factory
func RegisterAction(name string, factory ActionFactory) {
	// Register with normalized name
	normalized := normalizeActionName(name)
	actionRegistry[normalized] = factory
}

// GetAction creates an action instance by name
func GetAction(actionName string, sshService *services.SSHService, aptService *services.APTService, pctService *services.PCTService, containerID *string, cfg *libs.LabConfig, containerCfg *libs.ContainerConfig) (Action, error) {
	normalized := normalizeActionName(actionName)
	if factory, ok := actionRegistry[normalized]; ok {
		return factory(sshService, aptService, pctService, containerID, cfg, containerCfg), nil
	}
	return nil, fmt.Errorf("action '%s' not found. Available actions: %v", actionName, getAvailableActions())
}

func normalizeActionName(name string) string {
	// Normalize: lowercase, replace spaces/dashes/underscores with dashes
	name = strings.ToLower(name)
	name = strings.ReplaceAll(name, " ", "-")
	name = strings.ReplaceAll(name, "_", "-")
	return strings.TrimSpace(name)
}

func getAvailableActions() []string {
	actions := make([]string, 0, len(actionRegistry))
	for name := range actionRegistry {
		actions = append(actions, name)
	}
	return actions
}

File: csharp-enva/Actions/Registry.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Enva.Libs;
using Enva.Services;

namespace Enva.Actions;

public delegate IAction ActionFactory(SSHService? sshService, APTService? aptService, PCTService? pctService, string? containerID, LabConfig? cfg, ContainerConfig? containerCfg);

public static class ActionRegistry
{
    private static readonly Dictionary<string, ActionFactory> _actionRegistry = new();

    public static void RegisterAction(string name, ActionFactory factory)
    {
        string normalized = NormalizeActionName(name);
        _actionRegistry[normalized] = factory;
    }

    public static IAction? GetAction(string actionName, SSHService? sshService, APTService? aptService, PCTService? pctService, string? containerID, LabConfig? cfg, ContainerConfig? containerCfg)
    {
        string normalized = NormalizeActionName(actionName);
        if (_actionRegistry.TryGetValue(normalized, out var factory))
        {
            var action = factory(sshService, aptService, pctService, containerID, cfg, containerCfg);
            if (action is BaseAction baseAction)
            {
                baseAction.ActionName = actionName;
            }
            return action;
        }
        throw new Exception($"action '{actionName}' not found. Available actions: [{string.Join(", ", GetAvailableActions())}]");
    }

    private static string NormalizeActionName(string name)
    {
        string normalized = name.ToLower();
        normalized = normalized.Replace(" ", "-");
        normalized = normalized.Replace("_", "-");
        return normalized.Trim();
    }

    private static List<string> GetAvailableActions()
    {
        return _actionRegistry.Keys.ToList();
    }
}
